examples/cpp/waku.cpp

```diff
@@ -0,0 +1,336 @@
+#include <signal.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <stdint.h>
+#include <vector>
+#include <iostream>
+
+#include <sys/types.h>
+#include <unistd.h>
+#include <sys/syscall.h>
+
+#include "../../library/libwaku.h"
+
+#define WAKU_CALL(call)                                                        \
+do {                                                                           \
+  int ret = call;                                                              \
+  if (ret != 0) {                                                              \
+    std::cout << "Failed the call to: " << #call << ". Code: " << ret << "\n"; \
+  }                                                                            \
+} while (0)
+
+struct ConfigNode {
+    char    host[128];
+    int          port;
+    char     key[128];
+    int         relay;
+    char  peers[2048];
+};
+
+// Arguments parsing
+static char doc[] = "\nC example that shows how to use the waku library.";
+static char args_doc[] = "";
+
+static struct argp_option options[] = {
+    { "host",  'h', "HOST",  0, "IP to listen for for LibP2P traffic. (default: \"0.0.0.0\")"},
+    { "port",  'p', "PORT",  0, "TCP listening port. (default: \"60000\")"},
+    { "key",   'k', "KEY",   0, "P2P node private key as 64 char hex string."},
+    { "relay", 'r', "RELAY", 0, "Enable relay protocol: 1 or 0. (default: 1)"},
+    { "peers", 'a', "PEERS", 0, "Comma-separated list of peer-multiaddress to connect\
+ to. (default: \"\") e.g. \"/ip4/127.0.0.1/tcp/60001/p2p/16Uiu2HAmVFXtAfSj4EiR7mL2KvL4EE2wztuQgUSBoj2Jx2KeXFLN\""},
+    { 0 }
+};
+
+static error_t parse_opt(int key, char *arg, struct argp_state *state) {
+
+    struct ConfigNode *cfgNode = (ConfigNode *) state->input;
+    switch (key) {
+        case 'h':
+            snprintf(cfgNode->host, 128, "%s", arg);
+            break;
+        case 'p':
+            cfgNode->port = atoi(arg);
+            break;
+        case 'k':
+            snprintf(cfgNode->key, 128, "%s", arg);
+            break;
+        case 'r':
+            cfgNode->relay = atoi(arg);
+            break;
+        case 'a':
+            snprintf(cfgNode->peers, 2048, "%s", arg);
+            break;
+        case ARGP_KEY_ARG:
+            if (state->arg_num >= 1) /* Too many arguments. */
+            argp_usage(state);
+            break;
+        case ARGP_KEY_END:
+            break;
+        default:
+            return ARGP_ERR_UNKNOWN;
+        }
+
+    return 0;
+}
+
+static struct argp argp = { options, parse_opt, args_doc, doc, 0, 0, 0 };
+
+// Base64 encoding
+// source: https://nachtimwald.com/2017/11/18/base64-encode-and-decode-in-c/
+size_t b64_encoded_size(size_t inlen)
+{
+	size_t ret;
+
+	ret = inlen;
+	if (inlen % 3 != 0)
+		ret += 3 - (inlen % 3);
+	ret /= 3;
+	ret *= 4;
+
+	return ret;
+}
+
+const char b64chars[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
+
+void b64_encode(char* in, size_t len, std::vector<char>& out)
```

Comment:
If C++11 we should simple return with std::vector<char>. Move semantics and temporary object optimization will solve it for free.